### ISSUE TITLE
Snapshot accounts hash mismatch

### ISSUE THREAD

**Issue by carllin** [solana-labs/solana#17135]
#### Problem
While running a validator lemond on master, encountered the panic:

```
[2021-05-10T02:11:26.665696217Z WARN  solana_runtime::accounts] verify_bank_hash failed: MismatchedAccountHash
thread 'main' panicked at 'Snapshot bank for slot 77652854 failed to verify', /home/carl/solana/runtime/src/snapshot_utils.rs:631:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/88f19c6dab716c6281af7602e30f413e809c5974/library/std/src/panicking.rs:493:5
   1: std::panicking::begin_panic_fmt
             at /rustc/88f19c6dab716c6281af7602e30f413e809c5974/library/std/src/panicking.rs:435:5
   2: solana_runtime::snapshot_utils::bank_from_archive
   3: solana_ledger::bank_forks_utils::load
   4: solana_core::validator::Validator::new
   5: solana_validator::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2021-05-10T02:11:26.718486408Z ERROR solana_metrics::metrics] datapoint: panic program="validator" thread="main" one=1i message="panicked at 'Snapshot bank for slot 77652854 failed to verify', /home/carl/solana/runtime/src/snapshot_utils.rs:631:9" location="/home/carl/solana/runtime/src/snapshot_utils.rs:631:9"
```

Logs/ledger/snapshot are all under /home/carl/solana/DebugConsensus/mb/snapshot_fail_verify on lemond

**Comment by carllin** [comment]
Update: 

Running master vs 1.6 with 
```
/home/carl/solana/target/release/solana-ledger-tool -l /home/carl/solana/DebugConsensus/mb/validator-ledger verify --allow-dead-slots --halt-at-slot 77685413
```

reliably repros the hash mismatch at slot `77685413`. Seems like there's a hash mismatch on the account, though the lamports match

```
Value mismatch for pubkey: 5KW2RMYEAwr38GXNMa3PPe22sa2rbFjb9DFjCgegiT4H, 
good_value: (4JLaqFhGjKzgdSbEE1sC5jzK2NUJxYrW2sSg7K6oNNJJ, 10613255280), 
bad_value: (77kqomcnod4VXLtGopZy6VZ5nqcGyA3D16G33TPEdb6M, 10613255280)
```

**Comment by ryoqun** [comment]
@carllin hey, this looks like this? https://github.com/solana-labs/solana/issues/17083 (CC: @jeffwashington ) Latest development for this case is that AccountsDB is at defense right now (https://github.com/solana-labs/solana/issues/17083#issuecomment-836963482) lol

**Comment by carllin** [comment]
> @carllin hey, this looks like this? #17083 (CC: @jeffwashington ) Latest development for this case is that AccountsDB is at defense right now ([#17083 (comment)](https://github.com/solana-labs/solana/issues/17083#issuecomment-836963482)) lol

Maybe the read-only cache change: https://github.com/solana-labs/solana/pull/16150. We could try reverting this and running the `verify` again on master

**Comment by jeffwashington** [comment]
fwiw, [this](https://github.com/solana-labs/solana/pull/17170) should disable the read only cache

**Comment by jackcmay** [comment]
Scribbler!

**Comment by jeffwashington** [comment]
Carl's problem is different. I was looking for a hash mismatch assert. Carl's reproducible instructions above produce a different bank hash for a future slot, so no assert shows up - just different data in the verify. In a validator, we would mismatch with the result from others, presumably.

**Comment by jeffwashington** [comment]
I bisected this and found what I believe to be the problematic change wrt the hash mismatch. It is always possible I made a mistake. This change at least makes sense with what carl observed was happening.
#15410
@Lichtso 

[this branch](https://github.com/jeffwashington/solana/tree/hash_mismatch_pass) works
[this branch](https://github.com/jeffwashington/solana/tree/hash_mismatch_fail) fails

problematic pubkey is what carl found above:
5KW2RMYEAwr38GXNMa3PPe22sa2rbFjb9DFjCgegiT4H

to reproduce:
this is on lemond

`cargo build --release && sudo /home/jwash/solana/target/release/solana-ledger-tool -l /home/carl/solana/DebugConsensus/mb/validator-ledger verify --allow-dead-slots --halt-at-slot 77685413 >& fail.txt`

`tail -f log.txt | grep "5KW2RMYEAwr38GXNMa3PPe22sa2rbFjb9DFjCgegiT4H\|panic"`
if you see hash 77kq... fail
if you see hash 4JLa... pass

**Comment by jeffwashington** [comment]
There are other bank hash mismatch errors and a possible bad hdd on this machine that I'll get back to working on once we get a handle on the hash mismatch error. I can reproduce the bank hash mismatch and possible bad hdd errors with a different ledger folder on lemond.

**Comment by jeffwashington** [comment]
@ryoqun fyi
Since we are tracking several problems at the same time.

**Comment by jeffwashington** [comment]
data as logged is different:
fail:``` 03000000a27da00400000000011567d55e87e116c2308e085b63ccc77550f18c99f5ca5bd7edd012fca8580eac7f454c460201010000000000000000000300f7```
pass:```
03000000a562a10400000000011567d55e87e116c2308e085b63ccc77550f18c99f5ca5bd7edd012fca8580eac7f454c460201010000000000000000000300f7```

**Comment by jeffwashington** [comment]
account on pass: ```
lamports
 10613255280 data.len
 1524765 owner
 BPFLoaderUpgradeab1e11111111111111111111111 executable
 false rent_epoch
 179 data
 03000000a562a10400000000011567d55e87e116c2308e085b63ccc77550f18c99f5ca5bd7edd012fca8580eac7f454c460201010000000000000000000300f7```

**Comment by jeffwashington** [comment]
@Lichtso has a [fix/workaround](https://github.com/Lichtso/solana/commits/fix/invoke_context_push_account_deps). I confirmed [his fix on top of my hash_mismatch_fail branch](https://github.com/jeffwashington/solana/commit/ff2fe1984851ebb1a10c3579af3d6033d9e8feed) gets the correct hash for the account.
However, when I ran alex's branch with the fix applied to master, I got an account mismatch hash when initially scanning stores that is consistent with what I've been chasing on this machine. The mismatch went away when I went back to hash_mismatch_fail with the fix. So, I can confirm alex's fix gets the correct hash. Alex will talk with @jackcmay about the fix itself.
I then returned to work on the hash mismatch after loading from the snapshot, and it is gone again. This hash mismatch issue is the theorized scribbler mentioned above and it has been intermittent but common on this machine. I am trying to track the scribbler while preserving the ability to reproduce the issue alex is working on until that issue is completed.
